app/controllers/openid_connect/authorization_controller.rb

```diff
@@ -53,17 +53,13 @@ def check_sp_handoff_bounced
 
     def confirm_user_is_authenticated_with_fresh_mfa
       bump_auth_count unless user_fully_authenticated?
-
-      unless user_fully_authenticated? && service_provider_mfa_policy.
-          auth_method_confirms_to_sp_request?
-        return confirm_two_factor_authenticated(request_id)
+      if !user_fully_authenticated?
+        redirect_to new_user_session_url
+      elsif remember_device_expired_for_sp?
+        redirect_to user_two_factor_authentication_url
```

Comment:
Couple thoughts:

- Maybe the device expired should take priority here (at least if checking second-factor), if it was running as the last thing in the method previously?
- Is `user_fully_authenticated?` still accurate with the new redirect and removal of the MFA policy method? Or should the user only be redirected to sign in if there's no sign-in at all (i.e. haven't entered username/password)

```suggestion
      if !user_signed_in?
        redirect_to new_user_session_url
      elsif remember_device_expired_for_sp?
        redirect_to user_two_factor_authentication_url
```

Reply:
Hmm, yeah, you might be right about that. I pushed the first half in f86ff64, but curious to see if any existing tests fail, and will dig into overlap between remember device + fully authenticated.

Reply:
I added back the `user_fully_authenticated?` in 0e95583, and later refactored in d72c270 to try to get even closer to the initial logic.